### PR TITLE
known-exception-handling

### PR DESCRIPTION
**Pull request by MrRedAmber**
Added support for throwing known exceptions derived from bad API calls.

### CODE REVIEW

**Comment by MrRedAmber** [comment]
the checks are releasing false positives but the code works.

File: minio/known_errors.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
+# Author: MrRedAmber (https://github.com/MrRedAmber)
```
Comment (harshavardhana) [inline minio/known_errors.py]: Explicitly authors are not added as per our policy, we need to retain `Apache` license header from other file with date 2017.

File: minio/known_errors.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+# Credits:
+# - http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.htmlRESTErrorResponses
+# - https://github.com/riquito (pointing out my mistakes)
```
Comment (harshavardhana) [inline minio/known_errors.py]: While this is a nice gesture, it is not advisable to mention a github link of a github user. Github maintains the history of reviews if someone needs to look back.

File: minio/api.py [hunk 3]
```diff
@@ -1851,30 +1860,32 @@ def _url_open(self, method, bucket_name=None, object_name=None,
...
-        if response.status != 200 and \
-           response.status != 204 and response.status != 206:
+        if response.status != 200 and response.status != 204 and response.status != 206:
             # Upon any response error invalidate the region cache
             # proactively for the bucket name.
             self._delete_bucket_region(bucket_name)
 
             # Populate response_error with error response.
-            response_error = ResponseError(response)
+            # response_error = ResponseError(response, method, bucket_name)
```
Comment (harshavardhana) [inline minio/api.py]: why is this commented out?

File: minio/api.py [hunk 4]
```diff
@@ -1851,30 +1860,32 @@ def _url_open(self, method, bucket_name=None, object_name=None,
             dump_http(method, url, fold_case_headers, response,
                       self._trace_output_stream)
 
-        if response.status != 200 and \
-           response.status != 204 and response.status != 206:
+        if response.status != 200 and response.status != 204 and response.status != 206:
```
Comment (harshavardhana) [inline minio/api.py]: Why not keep the was before, PEP indicates 80 character line.

File: minio/api.py [hunk 5]
```diff
@@ -1626,7 +1636,7 @@ def _stream_put_object(self, bucket_name, object_name,
             # Further verify if we have matching md5sum as well.
             previous_part = uploaded_parts.get(part_number, None)
             if (previous_part and previous_part.size == current_part_size and
-                previous_part.etag == md5_hex):
+                        previous_part.etag == md5_hex):
```
Comment (harshavardhana) [inline minio/api.py]: this tabbing is wrong.

File: minio/api.py [hunk 6]
```diff
@@ -1601,15 +1610,16 @@ def _stream_put_object(self, bucket_name, object_name,
...
+        total_parts_count, part_size, last_part_size = optimal_part_info(
+            content_size)
 
         # Iter over the uploaded parts.
         parts_iter = self._list_object_parts(bucket_name,
                                              object_name,
                                              upload_id)
 
         # save uploaded parts for verification.
-        uploaded_parts = { part.part_number: part for part in parts_iter }
```
Comment (harshavardhana) [inline minio/api.py]: The space was kept on purpose for readability reasons, since its a one liner..

File: minio/api.py [hunk 7]
```diff
@@ -1359,7 +1364,7 @@ def presigned_put_object(self, bucket_name, object_name,
         is_non_empty_string(object_name)
 
         if expires.total_seconds() < 1 or \
-           expires.total_seconds() > _SEVEN_DAYS_SECONDS:
+                        expires.total_seconds() > _SEVEN_DAYS_SECONDS:
```
Comment (harshavardhana) [inline minio/api.py]: tabbing.

File: minio/api.py [hunk 8]
```diff
@@ -1326,7 +1331,7 @@ def presigned_get_object(self, bucket_name, object_name,
         :return: Presigned url.
         """
         if expires.total_seconds() < 1 or \
-           expires.total_seconds() > _SEVEN_DAYS_SECONDS:
+                        expires.total_seconds() > _SEVEN_DAYS_SECONDS:
```
Comment (harshavardhana) [inline minio/api.py]: tabbing

File: minio/api.py [hunk 9]
```diff
@@ -1191,7 +1192,8 @@ def list_incomplete_uploads(self, bucket_name, prefix=None,
 
         return self._list_incomplete_uploads(bucket_name, prefix, recursive)
 
-    def _list_incomplete_uploads(self, bucket_name, prefix=None, recursive=False, is_aggregate_size=True):
+    def _list_incomplete_uploads(self, bucket_name, prefix=None,
```
Comment (harshavardhana) [inline minio/api.py]: All these changes are unrelated .. lets not do formatting in a change which has nothing to do with it.

File: minio/api.py [hunk 10]
```diff
@@ -141,7 +142,7 @@ def __init__(self, endpoint, access_key=None,
         self._trace_output_stream = None
 
         self._conn_timeout = urllib3.Timeout.DEFAULT_TIMEOUT if not timeout \
-                             else urllib3.Timeout(timeout)
```
Comment (harshavardhana) [inline minio/api.py]: tabbing..

**Comment by harshavardhana** [comment]
Merge this already by fixing all the review comments as well - @MrRedAmber - thanks for the PR.
